### PR TITLE
Add function to calculate CAN bit timing from baudrate

### PR DESCRIPTION
**Pull request by iabdalkader**
* Calc bit timing from baudrate if provided.
* Allows sampling point override.
* Tested on F4, F7, H7 Y/V.
You can pass the individual args as usual or do something like:
```python
CAN(2, CAN.NORMAL, baudrate=125_000, sampling_point=75)
>>> CAN kernel clock: 50000000 prescaler: 20 bs1: 14 bs1: 5
```
See:
http://www.bittiming.can-wiki.info/?CLK=50&ctype=bxCAN&SamplePoint=75&calc=1

Note: This makes it a lot easier to make CAN work between different MCUs with different clocks, prescalers etc... I was going to add it as a util function initially but it has to be called before the constructor to get the values, so passing a baudrate seems better. The args enum is used twice so it may need to be moved to global scope.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for this, in general I think it's a good idea to be able to specify baudrate instead of the prescaler, as you say it makes it much more convenient across different MCUs.

Some questions/comments:
- Probably `baudrate` is a good name for the argument, it matches that in UART and indicates that it's "symbols per second".  Other options would be `bitrate` and `freq`.
- `sampling_point` is the same as `bs1` just in different units, is that correct?  I'm just wondering if there's a better name for this, eg `sampling_point_percent` to indicate the units.
- There are 3 parameters setting the bit properties: prescaler, bs1, bs2.  So if only 2 input values (baudrate and sampling-point) are used to determine these then there may be many solutions for prescaler, bs1, bs2.  Is there some reason to pick certain solutions over other ones, eg smaller prescaler is better?

File: ports/stm32/pyb_can.c [hunk 1]
```diff
@@ -140,9 +140,57 @@ STATIC void pyb_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_ki
...
+        }
+    }
+    #else // F4 and F7 and assume other MCUs too.
+    // CAN1/CAN2/CAN3 on APB1 use GetPCLK1Freq, alternatively use the following:
+    // can_kern_clk = ((HSE_VALUE / osc_config.PLL.PLLM ) * osc_config.PLL.PLLN) /
+    //  (osc_config.PLL.PLLQ * clk_init.AHBCLKDivider * clk_init.APB1CLKDivider);
+    can_kern_clk = HAL_RCC_GetPCLK1Freq();
+    #endif
+    for (int brp = 1; brp < 512; brp++) {// These values work on all MCUs
+        for (int bs1 = 1; bs1 < 16; bs1++) {
```
Comment (dpgeorge) [inline ports/stm32/pyb_can.c:176]: why not range up to 1024?

File: ports/stm32/pyb_can.c [hunk 2]
```diff
@@ -140,9 +140,57 @@ STATIC void pyb_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_ki
...
+            for (int bs2 = 1; bs2 < 8; bs2++) {
```
Comment (dpgeorge) [inline ports/stm32/pyb_can.c:177]: why not range up to 16?

File: ports/stm32/pyb_can.c [hunk 3]
```diff
@@ -140,9 +140,57 @@ STATIC void pyb_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_ki
     }
 }
 
+STATIC void pyb_can_calc_bit_timing(mp_arg_val_t *args) {
+    enum { ARG_mode, ARG_extframe, ARG_prescaler, ARG_sjw, ARG_bs1, ARG_bs2, ARG_auto_restart, ARG_baudrate, ARG_sampling_point};
```
Comment (dpgeorge) [inline ports/stm32/pyb_can.c:144]: These should definitely be global, so they are not duplicated.

Alternatively, could split this into a stand-alone function `uint32_t pyb_can_get_source_freq(void)`, and then put the rest of the searching logic in `pyb_can_init_helper` itself so the enum can stay private in that function.

File: ports/stm32/pyb_can.c (hunk 2 above)
Comment (iabdalkader) [reply ports/stm32/pyb_can.c:177]: For the ranges on the F4, F7 bs1 and bs2 max are the same (16, 8) and prescaler max is 1024... on the H7 they go up to 256 and 128 and prescaler max 512. So 512, 16 and 8 work on all MCUs, I found out it either converges real quick or doesn't. I can make the ranges go up to the max on each MCU your choice.

File: ports/stm32/pyb_can.c (hunk 3 above)
Comment (iabdalkader) [reply ports/stm32/pyb_can.c:144]: I like the split to `get_source_freq` and the search logic in `init_helper` more, will do that.

File: ports/stm32/pyb_can.c (hunk 2 above)
Comment (iabdalkader) [reply ports/stm32/pyb_can.c:177]: Actually I think the higher ranges may be for CAN FD mode, the driver doesn't support that.

**Comment by iabdalkader** [comment]
> * Probably `baudrate` is a good name for the argument, it matches that in UART and indicates that it's "symbols per second".  Other options would be `bitrate` and `freq`.

I've seen both used I think they're interchangeable, but baudrate is more common, can switch to bitrate if you like.

> * `sampling_point` is the same as `bs1` just in different units, is that correct?  I'm just wondering if there's a better name for this, eg `sampling_point_percent` to indicate the units.

It's the position of the last bit sample with respect to the whole bit time... You don't know how many Tqs that will be, so it must be specified as percentage, and I think it's understood as such... a better name is `sample_point`.
image
source http://www.bittiming.can-wiki.info/

> * There are 3 parameters setting the bit properties: prescaler, bs1, bs2.  So if only 2 input values (baudrate and sampling-point) are used to determine these then there may be many solutions for prescaler, bs1, bs2.  Is there some reason to pick certain solutions over other ones, eg smaller prescaler is better?

Yes maybe but not too critical for classical CAN I think, I'm not really a CAN expert, but according to the bittiming wiki more time quanta (smaller prescaler) is recommended, which what this algorithm generates anyway as the outermost  loop is the prescaler (so it starts with the smallest prescaler possible, loops over all the bs1 bs2 values, then increments the prescaler).
If you generate any table you'll see this comment:

> Yellow background rows are settings with recommended values, with an bittime consisting of 16 time quanta tq (Number of time quanta). At the time this tool was first developed, this seems to be the best value. in these days where we talk about CAN FD, as much as possible time quanta should be used to construct an bit time.

File: ports/stm32/pyb_can.c (hunk 2 above)
Comment (dpgeorge) [reply ports/stm32/pyb_can.c]: Ok, thanks for the explanation, I think it's good the way it is now (can improve later if needed).

**Comment by dpgeorge** [comment]
> I've seen both used I think they're interchangeable, but baudrate is more common, can switch to bitrate if you like.

Let's stick with baudrate to match UART.

> a better name is `sample_point`

Ok, good.

> but according to the bittiming wiki more time quanta (smaller prescaler) is recommended, which what this algorithm generates anyway as the outermost loop is the prescaler

Great!

File: ports/stm32/pyb_can.c [hunk 4]
```diff
@@ -162,6 +196,32 @@ STATIC mp_obj_t pyb_can_init_helper(pyb_can_obj_t *self, size_t n_args, const mp
...
+                        args[ARG_bs1].u_int = bs1;
+                        args[ARG_bs2].u_int = bs2;
+                        args[ARG_prescaler].u_int = brp;
+                        timing_found = true;
+                    }
+                }
+            }
+        }
+        if (timing_found == false) {
+            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Invalid baudrate and/or sampling point!"));
```
Comment (dpgeorge) [inline ports/stm32/pyb_can.c:221]: please use lowercase letter to start, and no exclamation mark, eg "couldn't match baudrate and sample point"

**Comment by dpgeorge** [comment]
Can you please add docs for these two new arguments?

**Comment by iabdalkader** [comment]
> Can you please add docs for these two new arguments?

I tried, can you check ?

File: docs/library/pyb.CAN.rst [hunk 5]
```diff
@@ -67,6 +67,10 @@ Methods
      - *auto_restart* sets whether the controller will automatically try and restart
        communications after entering the bus-off state; if this is disabled then
        :meth:`~CAN.restart()` can be used to leave the bus-off state
+     - *baudrate* if a baudrate other than 0 is provided, this function will automatically try to
+       calculate a CAN bit-timing that satisfies both the baudrate and the desired *sample_point*.
```
Comment (dpgeorge) [inline docs/library/pyb.CAN.rst:71]: maybe be clear that prescaler/bs1/bs2 are overridden, eg `... try to calculate a CAN bit-timing (overriding *prescaler*, *bs1* and *bs2*) that satisfies ...`

Comment (iabdalkader) [reply docs/library/pyb.CAN.rst]: done.

**Comment by dpgeorge** [comment]
Thanks for updating, squashed and merged in 20f8ce19825d1a976f8d61e79e661e67509681bd

**Pull request closed without merging**